docs: update README to include ImgixProvider and domain info

luqven opened the pull request:
This PR updates `README.md` to include documentation for the new ✨ `<ImgixProvider>` HOC and `domain` prop.

sherwinski commented on README.md:
```diff
@@ -33,6 +33,7 @@
         + [Fixed Image Rendering (i.e. non-flexible)](#fixed-image-rendering-ie-non-flexible)
         + [Background Mode](#background-mode)
         + [Picture Support](#picture-support)
+        + [ImgixProvider Component ✨](#imgixprovider-component)
```
I think we should remove the ✨ for stylistic consistency
```suggestion
        + [ImgixProvider Component](#imgixprovider-component)
```

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
 
+#### ImgixProvider Component
+
+> New 🪄 ✨
```
Not sure if this line was intentional but I think we should steer away from labeling features as "new" or otherwise because they will remained tied to a certain point in time.

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
+
+> New 🪄 ✨
+
+The [`<ImgixProvider>`]() Higher Order Component (HOC), makes its [props](#props) available to any nested `<Imgix>` component in your React application.
```
Looks like a hyperlink is missing here, or we can do away with it altogether

sherwinski commented on README.md:
````diff
@@ -325,6 +326,191 @@ const commonProps = {
+```html
+<div class="gallery">
+  <img
+    src="https://assets.imgix.net/examples/pione.jpg?auto=format&ar=16%3A9&"
````
There is an extra `&` appended to each of these URLs which should be removed

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
+    ...
+  />
+  <img
+    src="https://sdk-test.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg?auto=format&ar=16%3A9&"
```
Same as above

sherwinski commented on README.md:
````diff
@@ -325,6 +326,191 @@ const commonProps = {
+</div>
+```
+
+You can take advantage of this behavior to use 2-step, or partial, URLs with the `<Imgix>` component. By defining the [`domain`](#domain--string-optional) prop on the Provider, it can be made accessible to all nested `<Imgix>` components.
````
I'm fine with referring to partial paths as "2-step" internally but I'm wondering if it could be potentially confusing for our users to have two terms refer to the same thing? This is based off of past experiences with terms like DPR and fixed-sized srcsets.

My 2¢ is to stick to partial or full path URLs only as it's more intuitive.

sherwinski commented on README.md:
````diff
@@ -325,6 +326,191 @@ const commonProps = {
+}
+```
+
+Both the `<Imgix>` components above will access to the `domain` prop from the provider and have their relative `src` paths resolve to the same domain. So that the generated HTML looks something like
````
```suggestion
Both the `<Imgix>` components above will access to the `domain` prop from the provider and have their relative `src` paths resolve to the same domain. So that the generated HTML looks something like:
```

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
+<div class="gallery">
+  <img src="https://assets.imgix.net/examples/pione.jpg" ... />
+  <img
+    src="https://assets.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg"
```
Extremely nitpicky, but just wanted to point out that this URL doesn't resolve. Might want to consider replacing it with something that does for demonstration purposes.

sherwinski commented on README.md:
````diff
@@ -325,6 +326,191 @@ const commonProps = {
+}
+```
+
+So that the generated HTML looks something like this
````
```suggestion
So that the generated HTML looks something like this:
```

sherwinski commented on README.md:
````diff
@@ -325,6 +326,191 @@ const commonProps = {
+</div>
+```
+
+You cans nest `ImgixProvider` components to ensure that different consumers have different props.
````
```suggestion
You can nest `ImgixProvider` components to ensure that different consumers have different props.
```

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
+
+You cans nest `ImgixProvider` components to ensure that different consumers have different props.
+
+For example to give `Imgix` components different props from `Picture` components, you can next an `ImgixProvider` inside of another one.
```
```suggestion
For example to give `Imgix` components different props from `Picture` components, you can nest an `ImgixProvider` inside of another one.
```

sherwinski commented on README.md:
```diff
@@ -325,6 +326,191 @@ const commonProps = {
+        >
+          <Picture>
+            {/* imgixParams prop is no longer defined here */}
+            <Source
+              width={100}
+              htmlAttributes={{ media: "(min-width: 768px)" }}
+            />
```
No change needed but just wanted to say [whoa](https://media.giphy.com/media/uPnKU86sFa2fm/giphy.gif)

sherwinski commented on README.md:
```diff
@@ -426,6 +612,16 @@ These props are shared among Imgix and Source Components
 
+##### domain :: string, optional
+
+Required only when using 2-step, partial, paths as `src` prop for a component. IE, if `src` is `"/images/myImage.jpg"`, then the `domain` prop needs to be defined.
```
Same comment here about split use of "2-step" and "partial" paths.

sherwinski commented on README.md:
```diff
@@ -363,17 +361,17 @@ So that the generated HTML looks something like
   />
   <img
-    src="https://sdk-test.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg?auto=format&ar=16%3A9&"
+    src="https://sdk-test.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg?auto=format&ar=16%3"
```
Looks like you trimmed off the `A9` as well, so this won't give us an encoded `:` anymore. Also, I just noticed that this example ([`8382f4a` (#799)](https://github.com/imgix/react-imgix/pull/799/commits/8382f4a5e4e32d53731deed5369a0effe5bac2e5#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R344)) needs `fit: crop` in order for the `ar` example to work.

sherwinski reviewed (changes requested):
Noticed one other tiny detail, but otherwise it's almost there.

sherwinski commented on README.md:
```diff
@@ -325,6 +326,189 @@ const commonProps = {
+{
+  /*... */
+}
+<ImgixProvider imgixParams={{ ar: "16:9" }}>
```
if we're using `ar`, we should include `fit: "crop"` for completeness

sherwinski commented on README.md:
```diff
@@ -325,6 +326,189 @@ const commonProps = {
+  return (
+    <div className="App">
+      <header className="App-header">
+        <ImgixProvider imgixParams={{ ar: "16:9" }}>
```
Same here

sherwinski commented on README.md:
````diff
@@ -325,6 +326,189 @@ const commonProps = {
+```html
+<div class="gallery">
+  <img
+    src="https://assets.imgix.net/examples/pione.jpg?auto=format&ar=16%3"
+    ...
+  />
+  <img
+    src="https://sdk-test.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg?auto=format&ar=16%3"
````
Looks like you trimmed off the `A9` as well, so this won't give us an encoded `:` anymore. Also, I just noticed that this example ([`8382f4a` (#799)](https://github.com/imgix/react-imgix/pull/799/commits/8382f4a5e4e32d53731deed5369a0effe5bac2e5#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R344)) needs `fit: crop` in order for the `ar` example to work.

```suggestion
    src="https://assets.imgix.net/examples/pione.jpg?auto=format&fit=crop&ar=16%3A9"
    ...
  />
  <img
    src="https://sdk-test.imgix.net/%D8%B3%D8%A7%D9%86%D8%AF%D9%88%DB%8C%DA%86.jpg?auto=format&fit=crop&ar=16%3A9"
```

luqven replied:
Thanks good cathc. My apols for that botched fix.

sherwinski commented on README.md:
````diff
@@ -325,6 +326,189 @@ const commonProps = {
+```html
+<div class="gallery">
+  <img
+    src="https://assets.imgix.net/examples/pione.jpg?auto=format&ar=16%3"
````
```suggestion
    src="https://assets.imgix.net/examples/pione.jpg?auto=format&fit=crop&ar=16%3A9"
```

luqven commented:
> Noticed one other tiny detail, but otherwise it's almost there.

Thanks for these! Addressed your comments. Going to go ahead and merge this in once CI passes 👌🏼

frederickfogerty reviewed (commented):
Nice 👏
